File: src/Mobile/ntk_note_ip_app/lib/core/map/osm_static_map.dart
class GeoMarker {
  const GeoMarker({required this.lat, required this.lon, this.label});

  final double lat;
  final double lon;
  final String? label;
}

abstract final class OsmStaticMap {
  static String imageUrl({
    required String apiBaseUrl,
    required double latitude,
    required double longitude,
    int width = 640,
    int height = 280,
    int zoom = 10,
  }) {
    final base = _normalizeBaseUrl(apiBaseUrl);
    final lat = latitude.toStringAsFixed(5);
    final lon = longitude.toStringAsFixed(5);
    return '$base/api/v1/OsmMap/GetStatic'
        '?latitude=$lat&longitude=$lon&width=$width&height=$height&zoom=$zoom';
  }

  static String? buildAggregateMapUrl(
    String apiBaseUrl,
    List<GeoMarker> markers, {
    int width = 640,
    int height = 300,
  }) {
    if (markers.isEmpty) {
      return null;
    }

    final zoom = markers.length == 1
        ? 8
        : markers.length <= 4
            ? 4
            : 2;
    final markerParam = markers
        .take(15)
        .map((m) => '${m.lat.toStringAsFixed(5)},${m.lon.toStringAsFixed(5)}')
        .join('|');
    final base = _normalizeBaseUrl(apiBaseUrl);
    final uri = Uri.parse('$base/api/v1/OsmMap/GetStatic').replace(
      queryParameters: {
        'markers': markerParam,
        'width': '$width',
        'height': '$height',
        'zoom': '$zoom',
      },
    );
    return uri.toString();
  }

  static String openUrl(double latitude, double longitude) {
    final lat = latitude.toStringAsFixed(5);
    final lon = longitude.toStringAsFixed(5);
    return 'https://www.openstreetmap.org/?mlat=$lat&mlon=$lon#map=12/$lat/$lon';
  }

  static String _normalizeBaseUrl(String apiBaseUrl) {
    return apiBaseUrl.endsWith('/')
        ? apiBaseUrl.substring(0, apiBaseUrl.length - 1)
        : apiBaseUrl;
  }
}

File: src/Web/ClientApp/src/app/core/osm-static-map.ts
import { apiV1Group } from './api-routes';

export interface GeoMarker {
  lat: number;
  lon: number;
}

function normalizeBaseUrl(apiBaseUrl: string): string {
  return apiBaseUrl.replace(/\/$/, '');
}

export function buildStaticMapUrl(
  apiBaseUrl: string,
  latitude: number,
  longitude: number,
  width = 640,
  height = 280,
  zoom = 10
): string {
  const base = normalizeBaseUrl(apiBaseUrl);
  const params = new URLSearchParams({
    latitude: latitude.toFixed(5),
    longitude: longitude.toFixed(5),
    width: String(width),
    height: String(height),
    zoom: String(zoom),
  });

  return `${base}${apiV1Group('OsmMap')}/GetStatic?${params.toString()}`;
}

export function buildAggregateStaticMapUrl(
  apiBaseUrl: string,
  markers: GeoMarker[],
  width = 640,
  height = 300
): string | null {
  if (markers.length === 0) {
    return null;
  }

  const zoom = markers.length === 1 ? 8 : markers.length <= 4 ? 4 : 2;
  const markerParam = markers
    .slice(0, 15)
    .map((marker) => `${marker.lat.toFixed(5)},${marker.lon.toFixed(5)}`)
    .join('|');

  const base = normalizeBaseUrl(apiBaseUrl);
  const params = new URLSearchParams({
    markers: markerParam,
    width: String(width),
    height: String(height),
    zoom: String(zoom),
  });

  return `${base}${apiV1Group('OsmMap')}/GetStatic?${params.toString()}`;
}

export function buildOpenStreetMapUrl(latitude: number, longitude: number): string {
  const lat = latitude.toFixed(5);
  const lon = longitude.toFixed(5);
  return `https://www.openstreetmap.org/?mlat=${lat}&mlon=${lon}#map=12/${lat}/${lon}`;
}
